core/task.h

```diff
@@ -33,55 +33,150 @@
+
+  struct GateBatchGreater {
+    bool operator()(std::pair<bess::IGate *, bess::PacketBatch *> left,
+                    std::pair<bess::IGate *, bess::PacketBatch *> right) const {
+      return left.first->priority() > right.first->priority();
+    }
+  };
+
+  // XXX Tasks needs to be non-const in workers/modules
+  mutable bess::utils::extended_priority_queue<
+      std::pair<bess::IGate *, bess::PacketBatch *>, GateBatchGreater>
+      igates_to_run_;  // A queue for IGates to run
+
+  mutable bess::IGate *next_gate_;  // Cache next module to run without merging
+  // Optimization for chain
+  mutable bess::PacketBatch
+      *next_batch_;  // Cache to run next batch with next module
+
+  mutable bess::PacketBatch
+      dead_batch_;  // A packet batch for storing packets to free
+
+  // Simple packet batch pool
+  mutable int pbatch_idx_;
+  mutable bess::PacketBatch *pbatch_;
+
+  mutable std::vector<bess::PacketBatch *> gate_batch_;
+
+  /* The current input gate index is not given as a function parameter.
+   * Modules should use get_igate() for access */
+  mutable gate_idx_t current_igate_;
+
  public:
   // When this task is scheduled it will execute 'm' with 'arg'.  When the
   // associated leaf is created/destroyed, 'module_task' will be updated.
-  Task(Module *m, void *arg) : module_(m), arg_(arg), c_(nullptr) {}
+  Task(Module *m, void *arg)
+      : module_(m),
+        arg_(arg),
+        c_(nullptr),
+        igates_to_run_(),
+        next_gate_(),
+        next_batch_(),
+        pbatch_idx_(),
+        pbatch_(
+            new bess::PacketBatch[MAX_PBATCH_CNT]),  // XXX Need to adjust size
+        gate_batch_(std::vector<bess::PacketBatch *>(64, 0)) {
+    dead_batch_.clear();
+  }
+
+  ~Task() { delete[] pbatch_; }
 
   // Called when the leaf that owns this task is destroyed.
   void Detach();
 
   // Called when the leaf that owns this task is created.
   void Attach(bess::LeafTrafficClass *c);
 
+  inline void AddToRun(bess::IGate *ig, bess::PacketBatch *batch) const {
+    if (next_gate_ == nullptr &&
+        !ig->mergeable()) {  // optimization for chained
+      next_gate_ = ig;
+      next_batch_ = batch;
+    } else {
+      bess::PacketBatch *ibatch = get_gate_batch(ig);
+      if (ibatch && (static_cast<size_t>(ibatch->cnt() + batch->cnt()) <
+                     bess::PacketBatch::kMaxBurst)) {
+        // merge two batches
+        get_gate_batch(ig)->add(batch);
+      }
+      // set the input as new batch
+      set_gate_batch(ig, batch);
+      igates_to_run_.push(std::make_pair(ig, batch));
+    }
+  }
+
+  // Do not track used/unsued for efficiency
+  bess::PacketBatch *AllocPacketBatch() const {
+    DCHECK_LT(pbatch_idx_, MAX_PBATCH_CNT);
+    bess::PacketBatch *batch = &pbatch_[pbatch_idx_++];
+    batch->clear();
+    return batch;
+  }
+
+  void UpdatePerGateBatch(uint32_t gate_cnt) const {
+    if (gate_batch_.capacity() < gate_cnt) {
+      gate_batch_.resize(gate_cnt, nullptr);
```

Comment:
Dumb question: do you really want `reserve` here, not `resize`?  I ask because you're checking `capacity` not `size`.